### PR TITLE
Add protected branches client

### PR DESCRIPTION
**Pull request by Codex04**
- Protect a branch  
- Unprotect a branch  
- Get a single protected branch  
- Get a list of protected branches  

GitLab Protected Branches API Documentation: https://docs.gitlab.com/ee/api/protected_branches.html#protect-repository-branches

### CODE REVIEW

File: NGitLab/Models/AccessLevel.cs [hunk 1]
```diff
@@ -19,5 +19,6 @@ public enum AccessLevel
         /// Only valid for groups.
         /// </summary>
         Owner = 50,
+        NoAccess = 0,
```
Comment (meziantou) [inline NGitLab/Models/AccessLevel.cs]: You should move it to the first position

Comment (meziantou) [reply NGitLab/Models/AccessLevel.cs]: [According to the documentation](https://docs.gitlab.com/ee/api/protected_branches.html) there is an `Admin = 60` mode

File: NGitLab/Impl/ProtectedBranchesClient.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+            _api = api;
+            _projectId = projectId;
+            _protectedBranchesUrl = $"{Project.Url}/{_projectId}/protected_branches";
+        }
+
+        public ProtectedBranch ProtectBranch(BranchProtect branchProtect)
+            => _api.Post().With(branchProtect).To<ProtectedBranch>(_protectedBranchesUrl);
+
+        public void UnprotectBranch(string branchName)
+            => _api.Delete().Execute($"{_protectedBranchesUrl}/{branchName}");
```
Comment (meziantou) [inline NGitLab/Impl/ProtectedBranchesClient.cs]:
```suggestion
            => _api.Delete().Execute($"{_protectedBranchesUrl}/{Uri.EscapeDataString(branchName)}");
```

File: NGitLab/Impl/ProtectedBranchesClient.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+        public ProtectedBranch GetProtectedBranch(string branchName)
+            => _api.Get().To<ProtectedBranch>($"{_protectedBranchesUrl}/{branchName}");
```
Comment (meziantou) [inline NGitLab/Impl/ProtectedBranchesClient.cs]:
```suggestion
            => _api.Get().To<ProtectedBranch>($"{_protectedBranchesUrl}/{Uri.EscapeDataString(branchName)}");
```

File: NGitLab/Impl/ProtectedBranchesClient.cs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+        public ProtectedBranch GetProtectedBranches(string search)
+        {
+            var searchParameter = string.IsNullOrWhiteSpace(search) ? string.Empty : $"/search={search}";
```
Comment (meziantou) [inline NGitLab/Impl/ProtectedBranchesClient.cs]: The `search` parameter is in the query string, not the path. Also the value must be escaped.
You should use  `Utils.AddParameter` to set it.

**Review by meziantou — CHANGES_REQUESTED** [review]

**Review by meziantou — APPROVED** [review]

File: NGitLab/IProtectedBranchesClient.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
+﻿using NGitLab.Models;
+
+namespace NGitLab
+{
+    public interface IProtectedBranchesClient
```
Comment (louis-z) [inline NGitLab/IProtectedBranchesClient.cs]: Please rename to `IProtectedBranchClient`. As per most other client interfaces, it should be named with a singular noun followed by `Client`.

Rename derived types and methods too accordingly.

File: NGitLab/Models/ProtectedBranch.cs [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    [DataContract]
+    public class ProtectedBranch
+    {
+        [DataMember(Name = "id")]
+        public long Id { get; set; }
+
+        [DataMember(Name = "name")]
+        public string Name { get; set; }
+
+        [DataMember(Name = "merge_acess_levels")]
```
Comment (louis-z) [inline NGitLab/Models/ProtectedBranch.cs]:
```suggestion
        [DataMember(Name = "merge_access_levels")]
```

**Review by louis-z — CHANGES_REQUESTED** [review]

File: NGitLab/Models/ProtectedBranch.cs (hunk 6 above)
Comment (meziantou) [reply NGitLab/Models/ProtectedBranch.cs]: You should add a test to ensure the property names are ok

File: NGitLab.Mock/Clients/GitLabClient.cs [hunk 7]
```diff
@@ -83,5 +83,10 @@ public IEventClient GetProjectEvents(int projectId)
         public IWikiClient GetWikiClient(int projectId) => new WikiClient(Context, projectId);
 
         public IProjectLevelApprovalRulesClient GetProjectLevelApprovalRulesClient(int projectId) => new ProjectLevelApprovalRulesClient(Context, projectId);
+
+        public IProtectedBranchClient GetProtectedBranchesClient(int projectId)
```
Comment (louis-z) [inline NGitLab.Mock/Clients/GitLabClient.cs]:
```suggestion
        public IProtectedBranchClient GetProtectedBranchClient(int projectId)
```

File: NGitLab/GitLabClient.cs [hunk 8]
```diff
@@ -191,5 +191,8 @@ public IProjectLevelApprovalRulesClient GetProjectLevelApprovalRulesClient(int p
         {
             return new ProjectLevelApprovalRulesClient(_api, projectId);
         }
+
+        public IProtectedBranchClient GetProtectedBranchesClient(int projectId)
```
Comment (louis-z) [inline NGitLab/GitLabClient.cs]:
```suggestion
        public IProtectedBranchClient GetProtectedBranchClient(int projectId)
```

File: NGitLab/IGitLabClient.cs [hunk 9]
```diff
@@ -79,5 +79,7 @@ public interface IGitLabClient
         IGroupVariableClient GetGroupVariableClient(int groupId);
 
         IProjectLevelApprovalRulesClient GetProjectLevelApprovalRulesClient(int projectId);
+
+        IProtectedBranchClient GetProtectedBranchesClient(int projectId);
```
Comment (louis-z) [inline NGitLab/IGitLabClient.cs]:
```suggestion
        IProtectedBranchClient GetProtectedBranchClient(int projectId);
```

File: NGitLab/IProtectedBranchClient.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public interface IProtectedBranchClient
+    {
+        ProtectedBranch ProtectBranch(BranchProtect branchProtect);
+
+        void UnprotectBranch(string branchName);
+
+        ProtectedBranch GetProtectedBranch(string branchName);
+
+        ProtectedBranch GetProtectedBranches(string search);
```
Comment (louis-z) [inline NGitLab/IProtectedBranchClient.cs]: Won't that call return an array of `ProtectedBranch`es?

**Review by louis-z — CHANGES_REQUESTED** [review]

File: NGitLab/IProtectedBranchClient.cs (hunk 10 above)
Comment (Codex04) [reply NGitLab/IProtectedBranchClient.cs]: Yep that's true. I changed it

**Review by louis-z — APPROVED** [review]

File: NGitLab.Tests/ProtectedBranchTests.cs [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+                    new AccessLevelInfo()
+                    {
+                        AccessLevel = AccessLevel.NoAccess,
+                        Description = "Example",
+                    },
+                },
+                CodeOwnerApprovalRequired = false,
+            };
+#pragma warning disable MA0009 // Add regex evaluation timeout
+            var expectedBranchProtectJson = Regex.Replace(
```
Comment (meziantou) [inline NGitLab.Tests/ProtectedBranchTests.cs]: Not sure why you use a regex. You can use string.replace or `SimpleJson.Serialize(SimpleJson.Deserialize(...))`

**Pull request merged**